src/BayesFilters/src/sigma_point.cpp

```diff
@@ -32,21 +53,263 @@ void bfl::sigma_point::unscented_weights(const unsigned int n, const double alph
 }
 
 
-MatrixXd bfl::sigma_point::sigma_point(const Gaussian& state, const double c)
+MatrixXd bfl::sigma_point::sigma_point(const GaussianMixture& state, const double c)
 {
-    JacobiSVD<MatrixXd> svd = state.covariance().jacobiSvd(ComputeThinU);
+    MatrixXd sigma_points(state.dim, ((state.dim * 2) + 1) * state.components);
 
-    MatrixXd A = svd.matrixU() * svd.singularValues().cwiseSqrt().asDiagonal();
+    for (std::size_t i = 0; i < state.components; i++)
+    {
+        JacobiSVD<MatrixXd> svd = state.covariance(i).jacobiSvd(ComputeThinU);
 
-    MatrixXd sigma_points(state.dim, (state.dim * 2) + 1);
+        MatrixXd A = svd.matrixU() * svd.singularValues().cwiseSqrt().asDiagonal();
 
-    sigma_points << VectorXd::Zero(state.dim), std::sqrt(c) * A, -std::sqrt(c) * A;
+        Ref<MatrixXd> sp = sigma_points.middleCols(((state.dim * 2) + 1) * i, ((state.dim * 2) + 1));
 
-    if (state.dim_linear > 0)
-        sigma_points.topRows(state.dim_linear).colwise() += state.mean().topRows(state.dim_linear);
+        sp << VectorXd::Zero(state.dim), std::sqrt(c) * A, -std::sqrt(c) * A;
 
-    if (state.dim_circular > 0)
-        sigma_points.bottomRows(state.dim_circular) = directional_add(sigma_points.bottomRows(state.dim_circular), state.mean().bottomRows(state.dim_circular));
+        if (state.dim_linear > 0)
+            sp.topRows(state.dim_linear).colwise() += state.mean(i).topRows(state.dim_linear);
+
+        if (state.dim_circular > 0)
+            sp.bottomRows(state.dim_circular) = directional_add(sigma_points.bottomRows(state.dim_circular), state.mean(i).bottomRows(state.dim_circular));
+    }
 
     return sigma_points;
 }
+
+
+std::tuple<bool, GaussianMixture, MatrixXd> bfl::sigma_point::unscented_transform
+(
+    const GaussianMixture& input,
+    const UTWeight& weight,
+    FunctionEvaluation function
+)
+{
+    /* Sample sigma points. */
+    MatrixXd input_sigma_points = sigma_point::sigma_point(input, weight.c);
+
+    /* Propagate sigma points */
+    Data fun_data;
+    bool valid_fun_data;
+    std::tie(valid_fun_data, fun_data) = function(input_sigma_points.cast<float>());
+
+    /* Stop here if function evaluation failed. */
+    if (!valid_fun_data)
+        return std::make_tuple(false, GaussianMixture(), MatrixXd(0, 0));
+
+    /* For now casting Data to MatrixXd. */
+    MatrixXd prop_sigma_points = (bfl::any::any_cast<MatrixXf&&>(std::move(fun_data))).cast<double>();
```

Comment:
What is the meaning of having `Data` as return type if we are always any-casting to `MatrixXf`?
Am I missing something? If this is to cover possible future implementation of data that differs from `MatrixXf`, then, for the time being, we should write it somewhere to keep track of this choice.

Reply:
I think this is something is worth discussing a bit.

My concern is the following: if a function type of the form
```
std::tuple<bool, MatrixXf>(const Eigen::Ref<const Eigen::MatrixXf>&)
```
is used, then, inside the lambda functions, the output of `MeasurementModel::predictedMeasure` must be converted to `MatrixXf` even if this is not the right type (especially if later one has to evaluate means and covariance).

By returning a `bfl::Data`, instead, a possible future implementation of the core method `bfl::sigma_point::unscented_transform` could delegate the evaluation of mean and covariance to a method that takes `bfl::Data` as input and does the right thing.